### ISSUE TITLE
Running swiftlint in a Bazel sandbox

### ISSUE THREAD

**Issue by cgrindel** [realm/SwiftLint#3736]
### New Issue Checklist

- [X] Updated SwiftLint to the latest version
- [X] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

I am trying to run SwiftLint inside a [Bazel](https://bazel.build/) build action. The important part to note is that Bazel build actions are run inside of a sandbox (e.g., minimal PATH, known environment variables). The SwiftLint utility is being built using Swift Package Manager as part of the build. When I execute the swiftlint binary from my non-sandboxed terminal, all works as expected. When I run it from the sandbox, I get the `Illegal instruction: 4` error. 

I have seen the previous issues which were fixed by selecting the correct Xcode installation:

```
xcode-select -s /Applications/Xcode.app/Contents/Developer/
```

As you can see from the error output below, the `xcode-select -p` appears to be pointing to the correct Xcode installation. So, my question is what other environmental cues does SwiftLint use to execute? Any thoughts on how to debug this?

##### Complete output when running SwiftLint, including the stack trace and command used

```
+ xcode-select -p
/Applications/Xcode.app/Contents/Developer
+ xcodebuild -version
Xcode 13.0
Build version 13A233
+ /private/var/tmp/_bazel_chuck/8ceed168b1bc6ef9b1efd8dd50e39485/sandbox/darwin-sandbox/43/execroot/simple_with_binary_example/bazel-out/darwin-fastbuild/bin/lint_test.runfiles/swift_utils/SwiftLint/swiftlint version
0.44.0
+ /private/var/tmp/_bazel_chuck/8ceed168b1bc6ef9b1efd8dd50e39485/sandbox/darwin-sandbox/43/execroot/simple_with_binary_example/bazel-out/darwin-fastbuild/bin/lint_test.runfiles/swift_utils/SwiftLint/swiftlint lint --path main.swift --no-cache --enable-all-rules
/private/var/tmp/_bazel_chuck/8ceed168b1bc6ef9b1efd8dd50e39485/sandbox/darwin-sandbox/43/execroot/simple_with_binary_example/bazel-out/darwin-fastbuild/bin/lint_test.runfiles/simple_with_binary_example/lint_test: line 43: 89688 Illegal instruction: 4  "${swiftlint}" lint --path main.swift --no-cache --enable-all-rules
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? `0.44.0`
* Installation method used (Homebrew, CocoaPods, building from source, etc)? `Built via Swift Package Manager`
```sh
$ swift --version
swift-driver version: 1.26.9 Apple Swift version 5.5 (swiftlang-1300.0.31.1 clang-1300.0.29.1)
Target: x86_64-apple-macosx11.0
```
* Paste your configuration file: `NA`
* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? `No`
  If so, paste their relative paths and respective contents.
* Which Xcode version are you using (check `xcodebuild -version`)?
```
Xcode 13.0
Build version 13A233
```
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
// This triggers a violation:
let foo = try! bar()
```

```
echo "let foo = try! bar()" | "${swiftlint}" lint --use-stdin --no-cache --enable-all-rules
Output:
/private/var/tmp/_bazel_chuck/8ceed168b1bc6ef9b1efd8dd50e39485/sandbox/darwin-sandbox/49/execroot/simple_with_binary_example/bazel-out/darwin-fastbuild/bin/lint_test.runfiles/simple_with_binary_example/lint_test: line 42: 89977 Done                    echo "let foo = try! bar()"
     89978 Illegal instruction: 4  | "${swiftlint}" lint --use-stdin --no-cache --enable-all-rules
```

**Comment by blevasseur-godaddy** [comment]
I'm seeing the exact same issue where if the project isn't pre-launched via Xcode and cached (on an Apple Silicon environment), building via the terminal is throwing an illegal instruction 4 error. Note that Xcode is running in an apple silicon via Rosetta environment for me.

**Comment by jpsim** [comment]
I can run SwiftLint with bazel just fine with the following `WORKSPACE` configuration in an otherwise empty directory:

```starlark
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "SwiftLint",
    build_file_content = """exports_files(["swiftlint"])""",
    sha256 = "d9e56231675e27a72d9f7a6561ccb707cfe27f1aa529fcc06329ee7f4333ac6f",
    url = "https://github.com/realm/SwiftLint/releases/download/0.45.1/portable_swiftlint.zip",
)
```

```shell
$ echo "let foo = try! bar()" | bazel run @SwiftLint//:swiftlint -- --use-stdin --enable-all-rules
INFO: Analyzed target @SwiftLint//:swiftlint (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
INFO: Elapsed time: 0.144s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
<nopath>:1:1: warning: Explicit ACL Violation: All declarations should specify Access Control Level keywords explicitly. (explicit_acl)
<nopath>:1:1: warning: Explicit Top Level ACL Violation: Top-level declarations should specify Access Control Level keywords explicitly. (explicit_top_level_acl)
<nopath>:1:11: error: Force Try Violation: Force tries should be avoided. (force_try)
<nopath>:1:5: warning: Prefixed Top-Level Constant Violation: Top-level constants should be prefixed by `k`. (prefixed_toplevel_constant)
Done linting! Found 4 violations, 1 serious in 1 file.
```

Can you suggest changes to my configuration or run command that would demonstrate the issue in a way that I can reproduce it?

**Comment by cgrindel** [comment]
@jpsim It looks like you are using a pre-built swiftlint. The issue that I was seeing is trying to run the output of a Swift Package Manager build for swiftlint in a sandbox (e.g. inside a test). I will see if I can dig up an example that demonstrates the issue.

**Comment by jpsim** [comment]
Got it. Sample projects are super helpful for exactly this reasons 😄

**Comment by jpsim** [comment]
Just following up on this, I'd be happy to investigate further with a sample project @cgrindel

**Comment by jpsim** [comment]
As of https://github.com/jpsim/SourceKitten/pull/728 there's a way to run SwiftLint on macOS without using XPC to communicate with SourceKit.

Can you please see if https://github.com/realm/SwiftLint/pull/3900 fixes the sandboxing issues you were seeing when you set `IN_PROCESS_SOURCEKIT=YES` in the environment?

**Comment by cgrindel** [comment]
I will give this a shot.

**Comment by jpsim** [comment]
I'm adding support for building with Bazel in #4038. I'm also building on top of Bazel support to provide the ability for users of SwiftLint to define their own private, native custom rules in #4039.

I'd appreciate if folks who've commented on this issue could review those PRs and/or try them out in your projects. I'm interested in knowing if you run into any problems.

In my testing, I haven't run into any sandbox issues. Here's an example external `WORKSPACE`:

```python
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "build_bazel_rules_apple",
    sha256 = "36072d4f3614d309d6a703da0dfe48684ec4c65a89611aeb9590b45af7a3e592",
    url = "https://github.com/bazelbuild/rules_apple/releases/download/1.0.1/rules_apple.1.0.1.tar.gz",
)

load(
    "@build_bazel_rules_apple//apple:repositories.bzl",
    "apple_rules_dependencies",
)

apple_rules_dependencies()

load(
    "@build_bazel_rules_swift//swift:repositories.bzl",
    "swift_rules_dependencies",
)

swift_rules_dependencies()

load(
    "@build_bazel_rules_swift//swift:extras.bzl",
    "swift_rules_extra_dependencies",
)

swift_rules_extra_dependencies()

http_archive(
    name = "SwiftLint",
    sha256 = "a4321303f67ad0838d07cb59fe5831a39614182c544a06180789176ecdf13275",
    strip_prefix = "SwiftLint-2c5a6207958b317970b580b5b1c163540aad139b",
    url = "https://github.com/realm/SwiftLint/archive/2c5a6207958b317970b580b5b1c163540aad139b.tar.gz",
)

load("@SwiftLint//bazel:repos.bzl", "swiftlint_repos")

swiftlint_repos()

load("@SwiftLint//bazel:deps.bzl", "swiftlint_deps")

swiftlint_deps()
```

And I can run SwiftLint in the current working directory with `bazel run -c opt @SwiftLint//:swiftlint`

Let me know if this issue is resolved.
